File: 71/main.ts
function simplifyPath(path: string): string {
  let sections = generateSections(path);
  let stack: string[] = [];
  for (let section of sections) {
    if (section === "..") {
      stack.pop();
    } else if (section === ".") {
      continue;
    } else {
      stack.push(section);
    }
  }
  return "/" + stack.join("/");
}

function generateSections(path: string): string[] {
  let i = 0;
  let results: string[] = [];
  while (true) {
    while (path[i] === "/") i++;
    if (i >= path.length) break;
    let end = path.indexOf("/", i);
    if (end === -1) {
      results.push(path.slice(i));
      break;
    }
    results.push(path.slice(i, end));
    i = end;
  }
  return results;
}

// split by /\/+/
// iterate through words
// if ..
// pop stack
// if .
// continue
// otherwise
// push
// return  stack.join("/")

console.log(simplifyPath("/home/user/Documents/../Pictures"));

File: 71/main.py
class Solution:
    def simplifyPath(self, path: str) -> str:
        """
        approach
        use stack
        if "." do nothing
        if ".." pop
        if "/" push if not "/" on top, else, do nothing
        """
        stack = ["/"]
        i = 1
        n = len(path)

        def get_next_part(i: int):
            if i >= n:
                return None
            if path[i] == "/":
                i += 1
                return (i, "/")
            else:
                j = i
                while j < n and path[j] != "/":
                    j += 1
                return (j, path[i:j])

        while i < n:
            i, part = get_next_part(i)
            if part == ".":
                continue
            elif part == "..":
                if len(stack) < 3:
                    continue
                stack.pop()
                stack.pop()
                continue
            elif part == "/":
                if stack[-1] == "/":
                    continue
            stack.append(part)
        if len(stack) != 1 and stack[-1] == "/":
            stack.pop()
        return "".join(stack)
